File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* nodesBetweenCriticalPoints(struct ListNode* head, int* returnSize) {
    int* result = (int*)malloc(2 * sizeof(int));
    *returnSize = 2;
    result[0] = -1;
    result[1] = -1;
    int minDistance = INT_MAX;

    struct ListNode* previousNode = head;
    struct ListNode* currentNode = head->next;
    int currentIndex = 1;
    int pCIndex = 0;
    int fCIndex = 0;

    while(currentNode->next != NULL){
        if((currentNode->val < previousNode->val && currentNode->val <currentNode->next->val) || (currentNode->val > previousNode->val && currentNode->val > currentNode->next->val)){
            if(pCIndex == 0){
                pCIndex = currentIndex;
                fCIndex = currentIndex;
            }else{
                minDistance = fmin(minDistance , currentIndex - pCIndex);
                pCIndex = currentIndex;
            }
        }
        currentIndex++;
        previousNode = currentNode;
        currentNode = currentNode->next;
    }

    if(pCIndex != fCIndex){
        result[0] = minDistance;
        result[1] = pCIndex - fCIndex;
    }
    return result;
}

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        result = [-1 , -1]
        min_distance = float("inf")
        previous_node = head
        current_node = head.next
        current_index = 1
        pCIndex = 0
        fCIndex = 0

        while current_node.next is not None:
            if(current_node.val < previous_node.val and current_node.val < current_node.next.val) or (current_node.val > previous_node.val and current_node.val > current_node.next.val):
                if pCIndex == 0:
                    pCIndex = current_index
                    fCIndex = current_index
                else:
                    min_distance = min(min_distance , current_index - pCIndex)
                    pCIndex = current_index

            current_index += 1
            previous_node = current_node
            current_node = current_node.next
        if min_distance != float("inf"):
            max_distance = pCIndex - fCIndex
            result = [min_distance , max_distance]
        return result